### PR TITLE
CP-1664: modify prometheus-cpp

### PR DESCRIPTION
**Pull request by fcimeson**
The prometheus-cpp lib provides functionality to expose internal data to Prometheus via a series of sockets (unique to each node). Typically, each metric will get exposed for the lifetime of the process (node). Thus, if a node created 1000 metrics at the beginning of the program and never updated them again for the entire process, prometheus-cpp would continue exposing them to the socket for the entire process (similar to a while loop with 1000 print lines).

Additionally, prometheus-cpp uses a number of return by references, which poses the following issue:
```
auto& family = BuildCounter()....
auto& metric = family.Add(labels);
family.Remove(metric);
metric.Increment(); // Segfault
```
This can be fixed by returning a shared pointer instead. However, the nuance to this is that a shared pointer that has been removed may update its data but that data will not get exposed.

### CODE REVIEW

File: core/include/prometheus/counter.h [hunk 1]
```diff
@@ -47,7 +48,7 @@ class PROMETHEUS_CPP_CORE_EXPORT Counter {
   ClientMetric Collect() const;
 
  private:
-  Gauge gauge_{0.0};
```
Comment (fcimeson) [inline core/include/prometheus/counter.h:50]: The original code used other metrics to keep track of it's data. However, since the new code keeps track of its own retention time I have modified it to keep track of its own data to reduce duplication of the retention time variables.

File: core/include/prometheus/family.h [hunk 2]
```diff
@@ -22,6 +23,9 @@
 
 namespace prometheus {
 
+enum class RetentionBehavior {Keep, Remove};
```
Comment (fcimeson) [inline core/include/prometheus/family.h:29]: Originally, I had this as part of the class definition of Family but I wanted to avoid templating.

File: core/include/prometheus/family.h [hunk 3]
```diff
@@ -108,15 +114,15 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
   /// \return Return the newly created dimensional data or - if a same set of
   /// labels already exists - the already existing dimensional data.
   template <typename... Args>
-  T& Add(const std::map<std::string, std::string>& labels, Args&&... args) {
-    return Add(labels, detail::make_unique<T>(args...));
+  std::shared_ptr<T> Add(const std::map<std::string, std::string>& labels, Args&&... args) {
+    return Add(labels, std::make_shared<T>(std::forward<Args>(args)...));
   }
```
Comment (fcimeson) [inline core/include/prometheus/family.h:125]: This method is hear to avoid forward declarations of Args. It is less efficient since it creates an object that may or may not get used. We could instead move the definition here but that would be messy.

File: core/include/prometheus/metric_base.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  /// \brief    Check if the metric has expired
+  bool Expired() const;
+
+  template<typename T>
+  friend class Family;
+
+ protected:
+  bool has_family_{false};
+  std::atomic<double> retention_time_{1e9};
+  std::atomic<std::time_t> last_update_{std::time(nullptr)};
```
Comment (fcimeson) [inline core/include/prometheus/metric_base.h:44]: I'm not convinced we need these to be atomic, but it should be safer. Also, I tried to program it so that we could remove the atomic qualifier and the code would still compile.

File: core/src/histogram.cc [hunk 5]
```diff
@@ -8,45 +8,47 @@
...
   const auto bucket_index = static_cast<std::size_t>(std::distance(
       bucket_boundaries_.begin(),
       std::find_if(
           std::begin(bucket_boundaries_), std::end(bucket_boundaries_),
           [value](const double boundary) { return boundary >= value; })));
-  sum_.Increment(value);
-  bucket_counts_[bucket_index].Increment();
+  sum_ = sum_ + value;
+  bucket_counts_[bucket_index] = bucket_counts_[bucket_index] + 1;
+  last_update_ = std::time(nullptr);
```
Comment (fcimeson) [inline core/src/histogram.cc:25]: These three lines are atomic operations:
- sum_ = x; is equiv to sum_.store(x);
- sum_ is equiv to sum_.load()
- ...

The operators +=, -= didn't seem to be supported by my compiler. Also, the use of this syntax instead of load and store allows us to interchange atomic variables with non-atomic variables.

File: core/src/summary.cc [hunk 6]
```diff
@@ -3,18 +3,21 @@
...
+      quantile_values_(quantiles_, max_age, age_buckets) {}
+
+void Summary::Observe(const double value, const bool& alert) {
   std::lock_guard<std::mutex> lock(mutex_);
 
   count_ += 1;
   sum_ += value;
   quantile_values_.insert(value);
+  last_update_ = std::time(nullptr);
+  if (alert) AlertIfNoFamily();
```
Comment (fcimeson) [inline core/src/summary.cc]: AlertIfNoFamily => PrintAlerts

File: core/tests/counter_test.cc [hunk 7]
```diff
@@ -12,28 +12,28 @@ TEST(CounterTest, initialize_with_zero) {
 
 TEST(CounterTest, inc) {
   Counter counter;
-  counter.Increment();
+  counter.Increment(1, false);
```
Comment (fcimeson) [inline core/tests/counter_test.cc]: Can't test default and suppress alerts. Could, if I brought back the Increment method with no arguments.

File: core/include/prometheus/registry.h [hunk 8]
```diff
@@ -73,25 +74,29 @@ class PROMETHEUS_CPP_CORE_EXPORT Registry : public Collectable {
...
 
   const InsertBehavior insert_behavior_;
-  std::vector<std::unique_ptr<Family<Counter>>> counters_;
-  std::vector<std::unique_ptr<Family<Gauge>>> gauges_;
-  std::vector<std::unique_ptr<Family<Histogram>>> histograms_;
-  std::vector<std::unique_ptr<Family<Summary>>> summaries_;
+  std::vector<std::shared_ptr<Family<Counter>>> counters_;
+  std::vector<std::shared_ptr<Family<Gauge>>> gauges_;
+  std::vector<std::shared_ptr<Family<Histogram>>> histograms_;
+  std::vector<std::shared_ptr<Family<Summary>>> summaries_;
```
Comment (fcimeson) [inline core/include/prometheus/registry.h:99]: Maybe these should change to maps, to mock the behaviour of metrics in a family.

File: core/include/prometheus/family.h [hunk 9]
```diff
@@ -135,19 +141,21 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
   /// \return Zero or more samples for each dimensional data.
   std::vector<MetricFamily> Collect() override;
 
+  bool UpdateRetentionTime(const double& retention_time, const std::string& re_name, const std::map<std::string, std::string>& re_labels, const bool& bump = true, const bool& debug = false);
```
Comment (sarahttan) [inline core/include/prometheus/family.h]:
```suggestion
  bool UpdateRetentionTime(const double& retention_time, const std::string& re_name, 
                           const std::map<std::string, std::string>& re_labels, 
                           const bool& bump = true, const bool& debug = false);
```

File: core/src/summary.cc (hunk 6 above)
Comment (sarahttan) [reply core/src/summary.cc]: What does it mean if the alert has no family?  Is it that the metric/pointer is scheduled for removal?

File: core/include/prometheus/histogram.h [hunk 10]
```diff
@@ -67,8 +70,8 @@ class PROMETHEUS_CPP_CORE_EXPORT Histogram {
 
  private:
   const BucketBoundaries bucket_boundaries_;
-  std::vector<Counter> bucket_counts_;
-  Counter sum_;
+  std::atomic<double> sum_{0.0};
+  std::vector<std::atomic<double>> bucket_counts_;
```
Comment (sarahttan) [inline core/include/prometheus/histogram.h:74]: Why was this modified?

File: core/include/prometheus/metric_base.h (hunk 4 above)
Comment (sarahttan) [reply core/include/prometheus/metric_base.h:44]: I agree with you on this one.  I think that atomic is worth having for these variables.

File: core/src/family.cc [hunk 11]
```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
...
+  }
+  if (family.metric.empty()) {
+    return {};
+  } else {
+    return {family};
+  }
+}
+
+template <typename T>
+bool Family<T>::UpdateRetentionTime(const double& retention_time, const std::string& re_name, const std::map<std::string, std::string>& re_labels, const bool& bump, const bool& debug) {
```
Comment (sarahttan) [inline core/src/family.cc]:
```suggestion
bool Family<T>::UpdateRetentionTime(const double& retention_time, const std::string& re_name, 
                                    const std::map<std::string, std::string>& re_labels, 
                                    const bool& bump, const bool& debug) {
```

File: core/src/family.cc [hunk 12]
```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
...
+            break;
+          }
+        }
+        /* Update matched status (matched = true if all labels up to this point are a match) */
+        if (found) {
+          matched = true;
+        } else {
+          matched = false;
+          break;
+        }
```
Comment (sarahttan) [inline core/src/family.cc]:
```suggestion
        matched = found;
        if (!matched) break;
```

File: core/src/family.cc [hunk 13]
```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
...
+  /* Setup */
+  bool modified(false);
+  const boost::regex name_expr(re_name);
+  if (debug) {  
+    printf("UpdateRetentionTime: %s \n", re_name.c_str());
+  }
+  /* Check if the name is a match */
+  if (boost::regex_match(name_, name_expr)) {
+    for (auto metric: metrics_) {
+      bool matched(true);
```
Comment (sarahttan) [inline core/src/family.cc:115]: Based on this logic, I'm not too sure why we have both `matched` and `found`.  I think it looks like you could just use `matched` and then set it to true  under the constant or metric label conditions.

File: core/src/gauge.cc [hunk 14]
```diff
@@ -1,35 +1,30 @@
...
 
-void Gauge::Increment() { Increment(1.0); }
-
-void Gauge::Increment(const double value) {
-  if (value < 0.0) {
-    return;
-  }
-  Change(value);
+void Gauge::Increment(const double& value, const bool& alert) {
+  if (value < 0.0) return;
```
Comment (sarahttan) [inline core/src/gauge.cc:11]: While I think this line and the `Decrement` check makes sense, I also wonder if this check is necessary.  Like if the user wants chooses to decrement by a negative number, there won't be any warning or error raised, so would it be better to allow the user to do so?

File: core/src/summary.cc [hunk 15]
```diff
@@ -3,18 +3,21 @@
...
 Summary::Summary(const Quantiles& quantiles,
-                 const std::chrono::milliseconds max_age, const int age_buckets)
-    : quantiles_{quantiles},
-      count_{0},
-      sum_{0},
-      quantile_values_{quantiles_, max_age, age_buckets} {}
-
-void Summary::Observe(const double value) {
+                 const std::chrono::milliseconds& max_age,
+                 const int& age_buckets)
```
Comment (sarahttan) [inline core/src/summary.cc]: Not sure if this change is necessary (to minimize modifications) --> in particular `const int& age_buckets` versus `const int age_buckets`

**Review by sarahttan — CHANGES_REQUESTED** [review]

File: core/include/prometheus/histogram.h (hunk 10 above)
Comment (fcimeson) [reply core/include/prometheus/histogram.h:74]: Previously histogram used a series of counters to keep track of it's buckets and counters used guages to keep track of their values. This nesting of data structures proved troublesome when we now want to keep track of retention time for each metric (we don't want each bucket to have it's own retention time, and each counter in each bucket have it's own retention time >_<)

File: core/src/family.cc (hunk 12 above)
Comment (fcimeson) [reply core/src/family.cc]: There is sort of a bug in this logic. It should only update matched if not found.

File: core/src/family.cc (hunk 13 above)
Comment (fcimeson) [reply core/src/family.cc:115]: There is a bit of a bug in the logic, see followup comment.

File: core/src/family.cc (hunk 12 above)
Comment (fcimeson) [inline core/src/family.cc]:
```suggestion
        if (!found) {
          matched = false;
          break;
        }
```

File: core/src/gauge.cc (hunk 14 above)
Comment (fcimeson) [reply core/src/gauge.cc:11]: I agree but I'm not sure that warrants changing the original behaviour or not (the more that we leave the same, the easier it will be for merge updates).

File: core/src/summary.cc (hunk 15 above)
Comment (fcimeson) [reply core/src/summary.cc]: You caught me. I can change it back but I'll wait to see what the other reviewers say.

File: core/src/summary.cc (hunk 6 above)
Comment (fcimeson) [reply core/src/summary.cc]: Well, clearly you wouldn't understand since your family is awesome!

Joking aside, prometheus keeps a list of metrics (families) for publishing to sockets (which eventually get added to the database). Now that we have the ability to expire metrics (remove them from their family), we want to alert the user if and when they are trying to use a metric that will not get published.

**Review by sarahttan — APPROVED** [review]
To the best of my knowledge, I've reviewed the changes, but it's definitely a bit hard to switch review sites ;)

File: core/src/metric_base.cc [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <ctime>
+#include <iostream>
+
+namespace prometheus {
+
+bool MetricBase::HasFamily() { return has_family_; }
+
+void MetricBase::AlertIfNoFamily() {
+  if (!has_family_) std::cerr << "prometheus-cpp:: This metric has no family" << std::endl;
```
Comment (tt-avidbots) [inline core/src/metric_base.cc]: Could you extend the message to explain the consequence of not having a family (i.e. not publishing the metric to sockets) as you elaborated in one of your comments? It's not immediately clear why this important especially for new users.

File: core/include/prometheus/counter.h [hunk 17]
```diff
@@ -23,20 +27,17 @@ namespace prometheus {
...
   Counter() = default;
 
-  /// \brief Increment the counter by 1.
-  void Increment();
-
   /// \brief Increment the counter by a given amount.
   ///
   /// The counter will not change if the given amount is negative.
-  void Increment(double);
+  void Increment(const double& = 1, const bool& alert = true);
```
Comment (tt-avidbots) [inline core/include/prometheus/counter.h]: Could this alert be set during construction of the metric so the user doesn't have to add this argument if it is false every time they call the Increment methods on Counters, Gauges etc? It seems a bit cumbersome to add during each update when this needs to only be set once. I think it will also make the user more intentional about the alerts during the creation of the metrics.

File: core/include/prometheus/family.h (hunk 2 above)
Comment (tt-avidbots) [inline core/include/prometheus/family.h:29]: Could you add a comment to explain the significance/behaviour of the retention i.e. keeps metric for specified retention time, removes when retention time is exceeded? I wasn't sure how the removal worked and add to search for it.

File: core/include/prometheus/family.h [hunk 18]
```diff
@@ -86,10 +91,11 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
   /// \param name Set the metric name.
   /// \param help Set an additional description.
   /// \param constant_labels Assign a set of key-value pairs (= labels) to the
```
Comment (tt-avidbots) [inline core/include/prometheus/family.h:96]: Add documentation for the retention_behavior param.

File: core/include/prometheus/family.h [hunk 19]
```diff
@@ -135,19 +141,23 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
   /// \return Zero or more samples for each dimensional data.
   std::vector<MetricFamily> Collect() override;
 
+  bool UpdateRetentionTime(const double& retention_time, const std::string& re_name,
```
Comment (tt-avidbots) [inline core/include/prometheus/family.h]: Do we need to make built-in types const ref? I don't see any benefit to this but if there is let me know.

File: core/include/prometheus/family.h [hunk 20]
```diff
@@ -135,19 +141,23 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
...
+  bool UpdateRetentionTime(const double& retention_time, const std::string& re_name, 
+                           const std::map<std::string, std::string>& re_labels, 
+                           const bool& bump = true, const bool& debug = false);
+
  private:
-  std::unordered_map<std::size_t, std::unique_ptr<T>> metrics_;
+  std::unordered_map<std::size_t, std::shared_ptr<T>> metrics_;
   std::unordered_map<std::size_t, std::map<std::string, std::string>> labels_;
-  std::unordered_map<T*, std::size_t> labels_reverse_lookup_;
+  std::unordered_map<std::shared_ptr<T>, std::size_t> labels_reverse_lookup_;
```
Comment (tt-avidbots) [inline core/include/prometheus/family.h:166]: In practice, do we anticipate multiple pointers to the labels?

File: core/include/prometheus/family.h [hunk 21]
```diff
@@ -135,19 +141,23 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
...
   const std::string name_;
   const std::string help_;
   const std::map<std::string, std::string> constant_labels_;
+  RetentionBehavior retention_behavior_;
   std::mutex mutex_;
 
-  ClientMetric CollectMetric(std::size_t hash, T* metric);
-  T& Add(const std::map<std::string, std::string>& labels,
-         std::unique_ptr<T> object);
+  ClientMetric CollectMetric(std::size_t hash, std::shared_ptr<T> metric);
```
Comment (tt-avidbots) [inline core/include/prometheus/family.h:174]: Add documentation for this method.

File: core/include/prometheus/metric_base.h [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+  /// \brief    Prints a message to stderr if the metric doesn't have a family
+  void AlertIfNoFamily();
+
+  /// \brief    Update the metric's retention time
+  /// \param    retention_time the value used to set retention_time_
+  /// \param    bump if true, also update the metric's last_update time
+  void UpdateRetentionTime(const double& retention_time, const bool& bump = true);
+
+  /// \brief    Check if the metric has expired
+  bool Expired() const;
```
Comment (tt-avidbots) [inline core/include/prometheus/metric_base.h]: Expired --> IsExpired()

File: core/include/prometheus/registry.h [hunk 23]
```diff
@@ -73,25 +74,29 @@ class PROMETHEUS_CPP_CORE_EXPORT Registry : public Collectable {
   /// \return Zero or more metrics and their samples.
   std::vector<MetricFamily> Collect() override;
 
+  bool UpdateRetentionTime(const double& retention_time, const std::string& re_name, const std::map<std::string, std::string>& re_labels, const std::set<MetricType>& families = {MetricType::Counter, MetricType::Gauge, MetricType::Summary, MetricType::Histogram}, const bool& bump = true, const bool& debug = false);
```
Comment (tt-avidbots) [inline core/include/prometheus/registry.h]: This line is too long. Also add documentation for the method.

File: core/include/prometheus/summary.h [hunk 24]
```diff
@@ -72,11 +75,11 @@ class PROMETHEUS_CPP_CORE_EXPORT Summary {
   /// effectively results in a complete reset of the summary each time max_age
   /// has passed. The default value is 5.
   Summary(const Quantiles& quantiles,
-          std::chrono::milliseconds max_age = std::chrono::seconds{60},
-          int age_buckets = 5);
+          const std::chrono::milliseconds& max_age = std::chrono::seconds{60},
+          const int& age_buckets = 5);
```
Comment (tt-avidbots) [inline core/include/prometheus/summary.h]: Again I don't see any optimizations/improvements to making the built-in types const ref

File: core/src/counter.cc [hunk 25]
```diff
@@ -2,11 +2,14 @@
...
-void Counter::Increment(const double val) { gauge_.Increment(val); }
+void Counter::Increment(const double& value, const bool& alert) {
+  if (value < 0.0) return;
+  value_ = value_ + value;
+  last_update_ = std::time(nullptr);
+  if (alert) AlertIfNoFamily();
+}
 
-double Counter::Value() const { return gauge_.Value(); }
+double Counter::Value() const { return value_; }
```
Comment (tt-avidbots) [inline core/src/counter.cc:16]: I like the change you've made here with the counters etc keeping track of their values.

File: core/src/family.cc [hunk 26]
```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
...
+  {
+    std::lock_guard<std::mutex> lock{mutex_};
+    family.name = name_;
+    family.help = help_;
+    family.type = T::metric_type;
+    for (const auto& m : metrics_) {
+      if (!m.second->Expired()) {
+        family.metric.push_back(std::move(CollectMetric(m.first, m.second)));
+      } else if (retention_behavior_ == RetentionBehavior::Remove) {
+        to_be_removed.push_back(m.second);
```
Comment (tt-avidbots) [inline core/src/family.cc:88]: I get that we don't want to hold the lock for a long time but if you're using unordered maps for the metrics and labels, then the removal should be constant time in the average case. If removal from the unordered maps approach linear time then I think a better hash function needs to be chosen. Hence, do we need to store the metrics to be removed and then remove them later?

File: core/include/prometheus/family.h (hunk 19 above)
Comment (tt-avidbots) [inline core/include/prometheus/family.h]: Also, add method documentation for "UpdateRetentionTime"

File: core/src/family.cc [hunk 27]
```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
...
+  }
+  if (family.metric.empty()) {
+    return {};
+  } else {
+    return {family};
+  }
+}
+
+template <typename T>
+bool Family<T>::UpdateRetentionTime(const double& retention_time, const std::string& re_name,
```
Comment (tt-avidbots) [inline core/src/family.cc]: Couldn't you just update the retention time for each metric internally every time the metric is updated (incremented, decremented etc.) and set a flag for removal every time the retention time is exceeded to circumvent the name/metric matching process? 

Even if it remains as is, shouldn't this be a method that's called internally by the registry? Should the user explicitly have to call this function if they're set the retention behavior for a metric to remove?

File: core/include/prometheus/family.h (hunk 19 above)
Comment (fcimeson) [reply core/include/prometheus/family.h]: bools, ints, doubles no. Strings can have benifit as the addresss space is smaller than the string.

File: core/src/family.cc (hunk 26 above)
Comment (fcimeson) [reply core/src/family.cc:88]: That could be a solution. The Remove method does a few things, so I think I was just playing it safe. Additionally, the Remove method also has a lock, so that would either need to be removed (and only called when the data is locked, which is problematic) or change it to a scoped lock, which would need some thought.

I like the idea of calling it in place but I'm leaning towards leaving it since I like to play it safe with locks.

File: core/src/family.cc (hunk 27 above)
Comment (fcimeson) [reply core/src/family.cc]: The rentention time is updated automatically, these are just for the developer to keep the metric alive if they are not updating the metric with new data. As for flagging them for removal, we want to flag them peroidically, otherwise we might get a metric that gets data once during the whole program and then never expires.

The current implementation has the metrics update the class member last_update_ instead of calling UpdateRetentionTime since that method is used to find the metric or group of metrics and update their retention time.

File: core/include/prometheus/family.h (hunk 20 above)
Comment (fcimeson) [reply core/include/prometheus/family.h:166]: Labels no, metrics yes. This is a map from a metric to a label. Intuatively, I would have had the labels attached to the metric but the library uses Families to group metrics by label sets.

File: core/include/prometheus/family.h (hunk 19 above)
Comment (tt-avidbots) [reply core/include/prometheus/family.h]: Yea I meant the ints, doubles etc except string.

**Review by sarahttan — APPROVED** [review]

**Review by tt-avidbots — APPROVED** [review]
